keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+
+This limitation surfaces in multi socket, as well as single socket multi NUMA systems (commonly present in AMD hardware), and might cause a significant performance degradation for latency-critical execution and high-throughput applications, in a case when the Topology Manager schedules a pod on non-adjacent numa nodes.
+
+
+### Goals
+
+* Enable TopologyManager to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
+* Preserve all other properties of all policies
+
+## Proposal
+
+### Risks and Mitigations
+
+| Risk                                             | Impact | Mitigation                                                                                                                                                                               |
+|--------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but resources allocation can be spread among NUMA nodes which are non-adjacent what can lead to higher latency. |
+
+## Design Details
+
+
+### Proposed Change
+ 
+We propose to
+- add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.
+- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.
+
+Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
+This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.
+
+To summarize properties of `prefer-closest-numa-nodes` policy:
+
+* Preserve all properties of `best-effort` and `restricted` policies
+* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.
+
+When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
+Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).
+
+We would need to extend the `MachineInfo` struct with Distances field which would be an array of uint64 which will describe the distance between a given node and other nodes in the system.
+This information can be read from sysfs:
```

Comment:
asking for better understanding myself (in general): is the kernel the best (only?) data source to assess NUMA distances?

Reply:
So the Numa distance can be read from ACPI SLIT (system locality information table), which means that it is basically hardcoded inside firmware. Kernel exposes this information via `sysfs`. I don't know about any other method of retrieving this info.

Reply:
Thanks, I've limited knowledge in this area and looking forward to learn more. From my perspective trusting the kernel here is perfectly fine.